src/Features/Core/Portable/UnusedReferences/UnusedReferencesService.cs

```diff
@@ -0,0 +1,132 @@
+
+            foreach (var referenceType in _processingOrder)
+            {
+                if (!referencesByType.ContainsKey(referenceType))
+                {
+                    continue;
+                }
+
+                usedAssemblyReferences = DetermineUnusedReferences(referencesByType[referenceType], usedAssemblyReferences, unusedReferences);
+            }
+
+            return unusedReferences.ToImmutableArray();
+        }
+
+        private static IEnumerable<string> DetermineUnusedReferences(IEnumerable<ReferenceInfo> references, IEnumerable<string> usedAssemblyReferences, ImmutableArray<ReferenceInfo>.Builder unusedReferences)
+        {
+            var usedAssemblyLookup = usedAssemblyReferences.ToImmutableHashSet();
+
+            // Determine which toplevel references have compilation assemblies in set of used assemblies.
+            var toplevelUsedReferences = references.Where(reference =>
+                reference.CompilationAssemblies.Any(assembly => usedAssemblyLookup.Contains(assembly)));
+            var toplevelUsedAssemblyReferences = toplevelUsedReferences.SelectMany(reference => reference.GetAllCompilationAssemblies());
+
+            // Remove all assemblies that are brought into the compilation from the toplevel used references. When
+            // determining transtively used references this will reduce false positives.
+            var remainingReferences = references.Except(toplevelUsedReferences);
+            var remainingUsedAssemblyReferences = usedAssemblyReferences.Except(toplevelUsedAssemblyReferences);
+            var remainingUsedAssemblyLookup = remainingUsedAssemblyReferences.ToImmutableHashSet();
+
+            // Determine which transtive references have compilation assemblies in the set of remaining used assemblies.
+            foreach (var reference in remainingReferences)
+            {
+                var allCompilationAssemblies = reference.GetAllCompilationAssemblies();
+                if (allCompilationAssemblies.IsEmpty)
+                {
+                    // Analyzer packages won't contribute any assemblies to the compilation but will
+                    // show up as a reference. We will consider them used.
+                    continue;
+                }
+
+                if (!allCompilationAssemblies.Any(assembly => remainingUsedAssemblyLookup.Contains(assembly)))
+                {
+                    // References that do not contribute a used compilation assembly are considered unsused.
+                    unusedReferences.Add(reference);
+                    continue;
+                }
+
+                remainingUsedAssemblyReferences = remainingUsedAssemblyReferences.Except(allCompilationAssemblies);
+                remainingUsedAssemblyLookup = remainingUsedAssemblyReferences.ToImmutableHashSet();
+            }
+
+            return remainingUsedAssemblyReferences;
```

Comment:
In general it makes me nervous that `IEnumerable`s are re-used and returned in this, I'm not really sure when this will actually be executed. Perhaps this could return something stronger, like the `remainingUsedAssemblyLookup`, to be a bit more explicit.

Reply:
Will do.